prog.go

```diff
@@ -565,37 +622,44 @@ var haveProgTestRun = internal.FeatureTest("BPF_PROG_TEST_RUN", "4.12", func() e
 	return err
 })
 
-func (p *Program) testRun(in []byte, repeat int, reset func()) (uint32, []byte, time.Duration, error) {
-	if uint(repeat) > math.MaxUint32 {
-		return 0, nil, 0, fmt.Errorf("repeat is too high")
+func (p *Program) testRun(opts *RunOptions) (uint32, time.Duration, error) {
+	if len(opts.Data) == 0 {
+		return 0, 0, fmt.Errorf("missing input")
 	}
 
-	if len(in) == 0 {
-		return 0, nil, 0, fmt.Errorf("missing input")
+	if uint(len(opts.Data)) > math.MaxUint32 {
+		return 0, 0, fmt.Errorf("input is too long")
 	}
 
-	if uint(len(in)) > math.MaxUint32 {
-		return 0, nil, 0, fmt.Errorf("input is too long")
+	if err := haveProgTestRun(); err != nil {
+		return 0, 0, err
 	}
 
-	if err := haveProgTestRun(); err != nil {
-		return 0, nil, 0, err
+	ctx := new(bytes.Buffer)
+	if opts.Context != nil {
+		if err := binary.Write(ctx, internal.NativeEndian, opts.Context); err != nil {
+			return 0, 0, fmt.Errorf("cannot serialize context: %v", err)
+		}
 	}
 
-	// Older kernels ignore the dataSizeOut argument when copying to user space.
-	// Combined with things like bpf_xdp_adjust_head() we don't really know what the final
-	// size will be. Hence we allocate an output buffer which we hope will always be large
-	// enough, and panic if the kernel wrote past the end of the allocation.
-	// See https://patchwork.ozlabs.org/cover/1006822/
-	out := make([]byte, len(in)+outputPad)
+	var ctxOut []byte
+	if opts.ContextOut != nil {
+		ctxOut = make([]byte, reflect.Indirect(reflect.ValueOf(opts.ContextOut)).Type().Size())
```

Comment:
Using reflection here is incorrect, since the size of a type in memory != the size of its encoded form. For example, go type may contain padding due to alignment which binary.Size will ignore.

```suggestion
		ctxOut = make([]byte, binary.Size(opts.ContextOut))
```